client/core/bot.go

```diff
@@ -0,0 +1,1744 @@
+//     2 x L commitment. If low balance restricts the maintenance of L lots on
+//     one side, allow the difference in lots to be added to the opposite side.
+//  6. Place orders, cancels first, then buys and sells.
+type makerBot struct {
+	*Core
+	pgmID uint64
+	base  *makerAsset
+	quote *makerAsset
+	// TODO: enable updating of market, or just grab it live when needed.
+	market *Market
+	log    dex.Logger
+	book   *orderbook.OrderBook
+
+	running uint32 // atomic
+	wg      sync.WaitGroup
+	die     context.CancelFunc
+
+	rebalanceRunning uint32 // atomic
+
+	programV atomic.Value // *MakerProgram
+
+	ordMtx sync.RWMutex
+	ords   map[order.OrderID]*Order
+
+	oracleRunning uint32 // atomic
+}
+
+func createMakerBot(ctx context.Context, c *Core, pgm *MakerProgram) (*makerBot, error) {
+	if err := validateProgram(pgm); err != nil {
+		return nil, err
+	}
+	dc, err := c.connectedDEX(pgm.Host)
+	if err != nil {
+		return nil, err
+	}
+	xcInfo := dc.exchangeInfo()
+	supportedAssets := c.assetMap()
+
+	if _, ok := supportedAssets[pgm.BaseID]; !ok {
+		return nil, fmt.Errorf("base asset %d (%s) not supported", pgm.BaseID, unbip(pgm.BaseID))
+	}
+	if _, ok := supportedAssets[pgm.QuoteID]; !ok {
+		return nil, fmt.Errorf("quote asset %d (%s) not supported", pgm.QuoteID, unbip(pgm.QuoteID))
+	}
+
+	baseWallet := c.WalletState(pgm.BaseID)
+	if baseWallet == nil {
+		return nil, fmt.Errorf("no wallet found for base asset %d -> %s", pgm.BaseID, unbip(pgm.BaseID))
+	}
+	quoteWallet := c.WalletState(pgm.QuoteID)
+	if quoteWallet == nil {
+		return nil, fmt.Errorf("no wallet found for quote asset %d -> %s", pgm.QuoteID, unbip(pgm.QuoteID))
+	}
+
+	mktName := marketName(pgm.BaseID, pgm.QuoteID)
+	mkt := xcInfo.Markets[mktName]
+	if mkt == nil {
+		return nil, fmt.Errorf("market %s not known at %s", mktName, pgm.Host)
+	}
+
+	base := &makerAsset{
+		SupportedAsset: supportedAssets[pgm.BaseID],
+	}
+	base.balanceV.Store(baseWallet.Balance)
+
+	quote := &makerAsset{
+		SupportedAsset: supportedAssets[pgm.QuoteID],
+	}
+	quote.balanceV.Store(quoteWallet.Balance)
+
+	m := &makerBot{
+		Core:   c,
+		base:   base,
+		quote:  quote,
+		market: mkt,
+		// oracle:    oracle,
+		log:  c.log.SubLogger("BOT").SubLogger(marketName(pgm.BaseID, pgm.QuoteID)),
+		ords: make(map[order.OrderID]*Order),
+	}
+	m.programV.Store(pgm)
+
+	// Get a rate now, because max buy won't have an oracle fallback.
+	var rate uint64
+	if mkt.SpotPrice != nil {
+		rate = mkt.SpotPrice.Rate
+	}
+	if rate == 0 {
+		price, err := m.syncOraclePrice(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to establish a starting price: %v", err)
+		}
+		rate = mkt.ConventionalRateToMsg(price)
+	}
+
+	var maxBuyLots, maxSellLots uint64
+	maxBuy, err := c.MaxBuy(pgm.Host, pgm.BaseID, pgm.QuoteID, rate)
+	if err == nil {
+		maxBuyLots = maxBuy.Swap.Lots
+	} else {
+		m.log.Errorf("Bot MaxBuy error: %v", err)
+	}
+	maxSell, err := c.MaxSell(pgm.Host, pgm.BaseID, pgm.QuoteID)
```

Comment:
I think this should be part of a larger effort to handle "inventory management". Essentially, bots need to have have limits to how much funds they can access. I think `Core` will unfortunately need to be involved and it might be a pita.

Reply:
Yeah I was thinking the same thing, definitely sounds like a pita.